File: trunk/quote/Projects/PluginModelIO/QuoteLoader.vb
﻿Imports System.Reflection

Imports Model
Imports Model.Quote

Imports DB.QuoteDataBaseTableAdapters
Imports DB.QuoteDataBase
Imports System.Windows.Forms

Public Class QuoteLoader

    Dim _PropertyLoader As New ObjectGenerator

    Public Function Load(ByVal id As Long) As Header

        System.Windows.Forms.Cursor.Current = Cursors.WaitCursor

        Dim adaptor As New DB.QuoteDataBaseTableAdapters._QuoteTableAdapter
        Dim table As New DB.QuoteDataBase._QuoteDataTable
        Dim q As New Header()

        Dim customerObj As Model.Template.Customer

        adaptor.FillByByQuoteID(table, CDbl(id))
        If table.Rows.Count > 0 Then
            Dim row As DB.QuoteDataBase._QuoteRow = table.Rows(0)
            customerObj = Me.LookupCustomer(row)

            Dim rfq As String = ""
            If Not row.IsRequestForQuoteNumberNull Then
                rfq = row.RequestForQuoteNumber
            End If
            Dim part As String = ""
            If Not row.IsPartNumberNull Then
                part = row.PartNumber
            End If
            Dim templateID As Long
            If Not row.IsTemplateIDNull Then
                templateID = row.TemplateID
            End If
            q = New Header(row.id, rfq, part, _
                row.Initials, row.CreatedDate, row.LastModifedDate)
            
            Dim Initials As String = ""
            If Not row.IsInitialsNull Then
                Initials = row.Initials
            End If
            Dim createdDate As DateTime
            If Not row.IsCreatedDateNull Then
                createdDate = row.CreatedDate
            End If
            Dim lastModDate As DateTime
            If Not row.IsLastModifedDateNull Then
                lastModDate = row.LastModifedDate
            End If

            q.PrimaryProperties.CommonCustomer = customerObj
            q.PrimaryProperties.CommonPartNumber = part
            q.PrimaryProperties.CommonRequestForQuoteNumber = rfq
            q.PrimaryProperties.CommonCreatedDate = createdDate
            q.PrimaryProperties.CommonLastModified = lastModDate
            q.PrimaryProperties.CommonInitials = Initials
            LoadComponents(q)

            Dim o1 = LoadProperties(id, _
              CommonSaver.COMPUTATION_PROPERTIES_ID, q.ComputationProperties)
            q.SetPublicComputationProperties(o1)
            Dim o2 = LoadProperties(id, _
              CommonSaver.OTHER_PROPERTIES_ID, q.OtherProperties)
            q.SetPublicOtherProperties(o2)
            Dim o4 = LoadProperties(id, _
                CommonSaver.NOTE_PROPERTIES_ID, q.NoteProperties)
            q.SetPublicNoteProperties(o4)

        End If

        System.Windows.Forms.Cursor.Current = Cursors.Default

        Return q
    End Function

    Public Function LookupCustomer(row As DB.QuoteDataBase._QuoteRow) As Model.Template.Customer

        Dim customer As String
        Dim customerID As Integer

        If row.IsCustomerIDNull Then
            If Not row.IsCustomerNameNull Then
                Dim temp As Model.Template.Customer
                temp = Model.Template.Customer.GetByName(row.CustomerName)
                If (Not temp Is Nothing) Then
                    customerID = temp.ID
                End If
            End If
        Else
            customerID = row.CustomerID
        End If
        
        If row.IsCustomerNameNull Then
            customer = ""
        Else
            customer = row.CustomerName
        End If

        Dim customerObj As New Model.Template.Customer
        customerObj.SetName(customer)
        customerObj.SetID(customerID)

        Return customerObj
    End Function

    Private Function LoadProperties(ByVal id As Integer, _
                    ByVal childId As Integer, _
                    ByVal obj As Object) _
                    As Object

        Dim adaptor As New DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter

        _PropertyLoader = New ObjectGenerator

        Dim table As _QuotePropertiesDataTable = _
            adaptor.GetDataByQuoteIDAndPropertyID(id, childId)

        For Each row As _QuotePropertiesRow In table.Rows
            AddNode(row)
        Next

        Dim o As New Object
        _PropertyLoader.BaseTypeName = obj.GetType.FullName
        o = _PropertyLoader.Generate()
        Return o
    End Function

    Private Sub AddNode(ByVal row As _QuotePropertiesRow)

        Dim node As New ObjectGenerator.PropertyInfo
        node.Name = row.PropertyName
        If Not row.IsPropertyStringValueNull Then
            node.TypeName = "System.String"
            node.Value = row.PropertyStringValue
        ElseIf Not row.IsPropertyDecimalValueNull Then
            node.TypeName = "System.Decimal"
            node.Value = row.PropertyDecimalValue
        ElseIf Not row.IsPropertyIntegerValueNull Then
            node.TypeName = "System.Int32"
            node.Value = row.PropertyIntegerValue
        ElseIf Not row.IsPropertyDateValueNull Then
            node.TypeName = "System.String"
            Dim dt As DateTime = row.PropertyDateValue
            If dt.Year > 1900 Then
                node.Value = row.PropertyDateValue.ToShortDateString
            Else
                node.Value = ""
            End If
        End If
        If Not row.IsPropertyCatagoryNull Then
            node.Category = row.PropertyCatagory
        End If
        If Not row.IsPropertyDescriptionNull Then
            node.Description = row.PropertyDescription
        End If
        _PropertyLoader.Add(node)
    End Sub

    Public Sub LoadComponents(ByVal q As Model.Common.Header)

        Dim adaptor As New _QuoteDetailTableAdapter
        Dim partAdaptor As New WireComponentSourceTableAdapter
        Dim wireAdaptor As New WireSourceTableAdapter
        Dim gageAdaptor As New GageTableAdapter
        Dim id As Integer = q.PrimaryProperties.CommonId
        Dim table As _QuoteDetailDataTable = adaptor.GetDataByQuoteID(id)
        For Each row As _QuoteDetailRow In table.Rows

            Dim temp As New TempObj
            CommonLoader.LoadProperties(id, row.id, temp)

            Dim product As New Model.Product( _
              row.ProductCode, _
              temp.Gage, _
              0, _
              0, _
              row.IsWire, _
              "", _
              0,
              "", _
              0, _
              0)

            Dim detail As Model.Quote.Detail = q.NewDetail(product)
            detail.Qty = row.Qty
            With detail
                .IsWire = row.IsWire
                If Not row.IsSourceIDNull Then .SourceID = row.SourceID
                .SequenceNumber = row.SequenceNumber
                .UnitOfMeasure = row.UOM
            End With

            Dim o1 As Object = LoadProperties(id, row.id, detail.QuoteDetailProperties)
            If Me._PropertyLoader.NameList.Contains("MachineTime") Then detail.MachineTime = o1.MachineTime
            If Me._PropertyLoader.NameList.Contains("UnitCost") Then detail.UnitCost = o1.UnitCost
            detail.SetProperties(o1)
        Next
    End Sub

    Public Class TempObj
        Public Property UnitOfMeasure As String
        Public Property Gage As String = ""
    End Class

End Class

File: trunk/quote/Projects/PluginModelIO/QuoteLoader.cs
namespace Model.IO
{
    using System;
    using System.Linq;
    using System.Reflection;
    using System.Windows.Forms;
    
    using DB;
    using DB.QuoteDataBaseTableAdapters;
    
    using Model.Quote;

    public class QuoteLoader
    {
        private ObjectGenerator propertyLoader = new ObjectGenerator();
        
        public static Model.Customer LookupCustomer(DB.QuoteDataBase._QuoteRow row)
        {
            string customer = null;
            int customerID = 0;

            if (row.IsCustomerIDNull()) 
            {
                if (!row.IsCustomerNameNull()) 
                {
                    Model.Customer temp = null;
                    temp = Model.IO.Misc.CustomerDB.GetByName(row.CustomerName);
                    if (temp != null) 
                    {
                        customerID = temp.Id;
                    }
                }
            } 
            else 
            {
                customerID = (int)row.CustomerID;
            }

            if (row.IsCustomerNameNull()) 
            {
                customer = string.Empty;
            } 
            else 
            {
                customer = row.CustomerName;
            }

            Model.Customer customerObj = new Model.Customer();
            customerObj.SetName(customer);
            customerObj.SetId(customerID);

            return customerObj;
        }

        public Header Load(int id)
        {
            System.Windows.Forms.Cursor.Current = Cursors.WaitCursor;

            DB.QuoteDataBaseTableAdapters._QuoteTableAdapter adaptor = 
                new DB.QuoteDataBaseTableAdapters._QuoteTableAdapter();
            DB.QuoteDataBase._QuoteDataTable table = new DB.QuoteDataBase._QuoteDataTable();
            Header q = new Header();

            Model.Customer customerObj = null;

            adaptor.FillByByQuoteID(table, (decimal)Convert.ToDouble(id));
            if (table.Rows.Count > 0) 
            {
                DB.QuoteDataBase._QuoteRow row = table.Rows[0] as DB.QuoteDataBase._QuoteRow;
                customerObj = LookupCustomer(row);

                string rfq = string.Empty;
                if (!row.IsRequestForQuoteNumberNull()) 
                {
                    rfq = row.RequestForQuoteNumber;
                }
                
                string part = string.Empty;
                if (!row.IsPartNumberNull()) 
                {
                    part = row.PartNumber;
                }
                
                int templateID = 0;
                if (!row.IsTemplateIDNull()) 
                {
                    templateID = row.TemplateID;
                }
                
                q = new Header(
                    (int)row.id,
                    rfq, 
                    part, 
                    row.Initials, 
                    row.CreatedDate, 
                    row.LastModifedDate, 
                    templateID);

                string initials = string.Empty;
                if (!row.IsInitialsNull()) 
                {
                    initials = row.Initials;
                }
                
                DateTime createdDate = default(DateTime);
                if (!row.IsCreatedDateNull()) 
                {
                    createdDate = row.CreatedDate;
                }
                
                DateTime lastModDate = default(DateTime);
                if (!row.IsLastModifedDateNull()) 
                {
                    lastModDate = row.LastModifedDate;
                }

                q.PrimaryProperties.CommonCustomer = customerObj;
                q.PrimaryProperties.CommonPartNumber = part;
                q.PrimaryProperties.CommonRequestForQuoteNumber = rfq;
                q.PrimaryProperties.CommonCreatedDate = createdDate;
                q.PrimaryProperties.CommonLastModified = lastModDate;
                q.PrimaryProperties.CommonInitials = initials;
                
                this.LoadComponents(q);

                Common.ComputationProperties o1 = this.LoadProperties(
                    id,
                    CommonSaver.ComputationPropertiesId, 
                    q.ComputationProperties) as Common.ComputationProperties;
                
                q.SetPublicComputationProperties(o1);
                
                Common.OtherProperties o2 = this.LoadProperties(
                    id, 
                    CommonSaver.OtherPropertiesId, 
                    q.OtherProperties) as Common.OtherProperties;
                q.SetPublicOtherProperties(o2);
                
                Common.NoteProperties o4 = this.LoadProperties(
                    id, 
                    CommonSaver.NotePropertiesId, 
                    q.NoteProperties) as Common.NoteProperties;
                q.SetPublicNoteProperties(o4);
            }

            System.Windows.Forms.Cursor.Current = Cursors.Default;

            return q;
        }

        public void LoadComponents(Model.Common.Header header)
        {
            _QuoteDetailTableAdapter adaptor = new _QuoteDetailTableAdapter();
            
            int id = header.PrimaryProperties.CommonId;
            DB.QuoteDataBase._QuoteDetailDataTable table = adaptor.GetDataByQuoteID(id);

            foreach (DB.QuoteDataBase._QuoteDetailRow row in table.Rows) 
            {
                QuoteLoaderTempObj temp = new QuoteLoaderTempObj();
                CommonLoader.LoadProperties(id, (int)row.id, temp);

                ProductBuildData data = new ProductBuildData();
                data.Code = row.ProductCode;
                data.Gage = temp.Gage;
                data.IsWire = row.IsWire;
                data.Vendor = string.Empty;
                data.Description = string.Empty;
                Product product = new Product(data);

                Model.Quote.Detail detail = header.NewDetail(product) as Model.Quote.Detail;
                
                detail.Qty = row.Qty;
                
                if (!row.IsSourceIDNull())
                {
                    detail.SourceId = row.SourceID;
                }
                
                detail.SequenceNumber = row.SequenceNumber;
                detail.UnitOfMeasure = row.UOM;

                Common.ISavableProperties o1 = null;
                if (detail.IsWire)
                {
                    o1 = this.LoadProperties(
                        id, 
                        (int)row.id,
                        detail.QuoteDetailProperties) 
                        as Common.ISavableProperties;
                }
                else
                {
                    o1 = this.LoadProperties(
                        id, 
                        (int)row.id,
                        detail.QuoteDetailProperties) 
                        as Common.ISavableProperties;
                }
                
                if (this.propertyLoader.NameList.Contains("MachineTime"))
                {
                    PropertyInfo prop = o1.GetType().GetProperty("MachineTime");
                    if (prop != null)
                    {
                        decimal value = (decimal)prop.GetValue(o1, new object[] { });
                        detail.MachineTime = value;
                    }
                }
                
                if (this.propertyLoader.NameList.Contains("UnitCost"))
                {
                    PropertyInfo prop = o1.GetType().GetProperty("UnitCost");
                    if (prop != null)
                    {
                        decimal value = (decimal)prop.GetValue(o1, new object[] { });
                        detail.UnitCost = value;
                    }
                }
                
                detail.SetProperties(o1);
            }
        }
        
        private object LoadProperties(
            int id, 
            int childId, 
            object obj,
            Type interfaceToImplement = null)
        {
            DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter adaptor = 
                new DB.QuoteDataBaseTableAdapters._QuotePropertiesTableAdapter();

            this.propertyLoader = new ObjectGenerator();
            
            if (interfaceToImplement != null)
            {
                this.propertyLoader.InterfaceToImplement = interfaceToImplement.FullName;
            }

            DB.QuoteDataBase._QuotePropertiesDataTable table = 
                adaptor.GetDataByQuoteIDAndPropertyID(id, childId);

            foreach (DB.QuoteDataBase._QuotePropertiesRow row in table.Rows) 
            {
                this.AddNode(row);
            }

            object o = new object();
            this.propertyLoader.BaseTypeName = obj.GetType().FullName;
            o = this.propertyLoader.Generate();
            
            return o;
        }

        private void AddNode(DB.QuoteDataBase._QuotePropertiesRow row)
        {
            ObjectGeneratorPropertyInfo node = new ObjectGeneratorPropertyInfo();
            node.Name = row.PropertyName;
            
            if (!row.IsPropertyStringValueNull()) 
            {
                node.TypeName = "System.String";
                node.Value = row.PropertyStringValue;
            } 
            else if (!row.IsPropertyDecimalValueNull()) 
            {
                node.TypeName = "System.Decimal";
                node.Value = row.PropertyDecimalValue;
            } 
            else if (!row.IsPropertyIntegerValueNull()) 
            {
                node.TypeName = "System.Int32";
                node.Value = row.PropertyIntegerValue;
            } 
            else if (!row.IsPropertyDateValueNull()) 
            {
                node.TypeName = "System.String";
                DateTime dt = row.PropertyDateValue;
                if (dt.Year > 1900) 
                {
                    node.Value = row.PropertyDateValue.ToShortDateString();
                } 
                else 
                {
                    node.Value = string.Empty;
                }
            }
            
            if (!row.IsPropertyCatagoryNull()) 
            {
                node.Category = row.PropertyCatagory;
            }
            
            if (!row.IsPropertyDescriptionNull()) 
            {
                node.Description = row.PropertyDescription;
            }
            
            this.propertyLoader.Add(node);
        }
    }
}